### PR TITLE
[BYOC] Allow custom codegens to register their own constant updater

### PR DESCRIPTION
**Pull request by mbaret**
Currently, all codegens using BYOC must make use of the default ConstantUpdater pass. However, certain codegens, like Ethos-N, don't want to store any constants in metadata module. This provides an interface (via a global) to register a custom constant updating method and assigns a 'null' updater for the Ethos-N codegen.

### CODE REVIEW

**Comment by mbaret** [comment]
cc @manupa-arm @zhiics @comaniac

File: src/relay/backend/contrib/codegen_c/codegen.cc [hunk 1]
```diff
@@ -290,8 +290,37 @@ runtime::Module CCompiler(const ObjectRef& ref) {
   return csource.CreateCSourceModule(ref);
 }
 
+/*!
+ * \brief A visitor to add the constants used as params for MetadataModule.
+ */
+struct CCompilerConstantUpdater : public ExprVisitor {
```
Comment (zhiics) [inline src/relay/backend/contrib/codegen_c/codegen.cc:296]: why is this needed?

Comment (mbaret) [reply src/relay/backend/contrib/codegen_c/codegen.cc:296]: Essentially just so that I can write a test case, because Ethos-N registers a 'null' constant updater. This also has the benefit of removing the implicit coupling between the CCompiler and the default ConstantUpdater (currently they must both adopt the same naming and visiting convention).

Comment (zhiics) [reply src/relay/backend/contrib/codegen_c/codegen.cc:296]: I think this doesn't couple to the updater though because it does nothing as what the constant updater does.

Comment (mbaret) [reply src/relay/backend/contrib/codegen_c/codegen.cc:296]: The concern is that the ConstantUpdater has no direct coupling with the external codegens, and yet it is generic. That is to say, the C Compiler codegen _must_ visit and name its constants in the same way that the ConstantUpdater chooses to, or you get a runtime failure. However, this relationship isn't enforced by any mechanism and relies on the codegen implementer to understand they must follow the convention. A simple example where this may happen is if a codegen only wants to save some constants and not others - the Ethos-N codegen being an extreme example where it wants to save none of them.

Comment (zhiics) [reply src/relay/backend/contrib/codegen_c/codegen.cc:296]: How about just add a sentence in the file description to explain that the constant updater directly uses the default/generic one so that we don't really need to have the exactly same one here? Since C compiler is not really for production use, I think we should keep is simple and clean.

Comment (manupa-arm) [reply src/relay/backend/contrib/codegen_c/codegen.cc:296]: @zhiics , What about the other way ? We make every codegen register a constant updater -- as you say if most of the codegen are based on JSONRuntime maybe that can implement the generic one and which could be overridden if the codegen requires. This way it creates a nice division of responsibilities -- constant creation and consumption is at more or less the same set of sources.

Comment (zhiics) [reply src/relay/backend/contrib/codegen_c/codegen.cc:296]: @manupa-arm I think one of the problems we are trying to avoid is adding more and more explicit registries to the external codegens. But it doesn't seems necessary to me to register a constant updater for each codegen because currently probably only ethos-n needs an empty updater. I think we can just fallback to generic one if there is no need for a special updater.

Comment (manupa-arm) [reply src/relay/backend/contrib/codegen_c/codegen.cc:296]: I do agree we need to avoid more and more explicit registries but this registry is more of a requirement to enable codegens based on other languages to be linked to TVM -- the same argument as we do relay.ext.<codegen_name> registry.

The problem I see with the current design of constant updating in this manner is if you look at the VisitExpr of ConstantNode, the correct lookup of constant is entirely dependent of the constant_idx that is generated by a generic constant updater living outside the codegen. This locks the down the design of external codegens to traverse in the same way GenericConstant updater is traversing and also blocks any modification to constant if such is required by the external codegen (e.g., compression, adding a header, etc.).

To highlight the further the issue of locking down the traversal will become more damaging when the graph becomes less-linear (the ones that could be generated via NASs in future). In which case we might have a degree of freedom of scheduling (i.e. we might want to schedule one operator before the other for various reasons -- such as caching, memory, performance that depends on the ext backend). 

Although, I appreciate that its makes the life quite easy of a tvm developer to have the generic constant updater but however I think it should live in the codegen itself. It also improves readabilty as well -- because now we can figure out the why the constants are indexed in the same way and how it retrieves the exact constant due to the fact the codegen post it to the params in the first place -- rather than having a generic constant updater living outside the codegen sources.

Maybe its outside of the scope for this PR -- but this PR exposed a symptom of the real problem, IMHO.

Comment (manupa-arm) [reply src/relay/backend/contrib/codegen_c/codegen.cc:296]: If you agree to what I say, I think we should slowly start showing this style of design in ccompiler and promote it to be used by external codegen developers to make it future proof and deprecate the generic one living outside the codegen once we have every codegen comply to it.

Comment (zhiics) [reply src/relay/backend/contrib/codegen_c/codegen.cc:296]: @manupa-arm Thanks for clarification. I am not arguing the necessity of constant updater for codegens. My point is that we can probably just add to the ones that needs it. To keep consistency, you can also register it for each of the other codegens but we don't really need to copy the code everywhere if there can fully reuse the generic one.

Comment (mbaret) [reply src/relay/backend/contrib/codegen_c/codegen.cc:296]: So the reason I added it for ccompiler is really just so that I could write a test for this code pathway. The Ethos-N updater doesn't really do anything so it's not great for testing. I could just temporarily register a new constant updater for ccompiler in the test instead, would that be better?

File: src/relay/backend/graph_runtime_codegen.cc [hunk 2]
```diff
@@ -371,10 +371,28 @@ class GraphRuntimeCodegen : public backend::MemoizedExprTranslator<std::vector<G
...
       // collect metadata.
+      auto codegen = func->GetAttr<String>(attr::kCompiler);
+      CHECK(codegen.defined()) << "No external codegen is set";
+      std::string codegen_name = codegen.value();
       const auto name_node = func->GetAttr<String>(tvm::attr::kGlobalSymbol);
-      std::string symobl = std::string(name_node.value());
-      ConstantUpdater const_visit(symobl, &params_);
-      const_visit(func);
+      std::string symbol = std::string(name_node.value());
+      std::string const_update_name = "relay.ext." + codegen_name + ".constant_updater";
```
[the remark this answers is not in the document]
Comment (mbaret) [reply src/relay/backend/graph_runtime_codegen.cc:379]: Could you elaborate on this? I'm not sure what's meant my moving it to the updater.

File: src/relay/backend/contrib/codegen_c/codegen.cc (hunk 1 above)
Comment (zhiics) [reply src/relay/backend/contrib/codegen_c/codegen.cc]: Yeah, that would be better. Thanks.

File: src/relay/backend/graph_runtime_codegen.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (zhiics) [reply src/relay/backend/graph_runtime_codegen.cc:379]: I meant that we could check the extern updaters in the constant updater itself, i.e. if there is one we use that one, otherwise we use the generic one. Then, we don't have to do have this in both graph_runtime_codegen.cc and compiler.cc.

[the remark this answers is not in the document]
Comment (mbaret) [reply src/relay/backend/graph_runtime_codegen.cc:379]: Do you think it might be cleaner to introduce a new function in backend/utils called something like UpdateConstants which replicates the logic that currently lives directly in graph runtime/vm? That way we can keep the generic ConstantUpdater clean.

[the remark this answers is not in the document]
Comment (zhiics) [reply src/relay/backend/graph_runtime_codegen.cc:379]: I think we probably don't need to though. As you can see, currently it is just one/two lines of code.

**Comment by mbaret** [comment]
It seems the custom updater I registered in the test persists between tests in CI and causes failures. Is there a way to unregister a function? @tqchen

**Comment by tqchen** [comment]
this is because pytest will run all testcases together, remove a global func exist in c++ API https://github.com/apache/incubator-tvm/blob/main/src/runtime/registry.cc#L77 we could expose that as a PackedFunc in tvm._ffi.remove_global_func(name)

File: src/relay/backend/graph_runtime_codegen.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (mbaret) [reply src/relay/backend/graph_runtime_codegen.cc:379]: I'm not exactly sure what you mean - these two lines specifically get the symbol from the name_node and the name of the constant updater global func. You need the symbol to initialise the ConstantUpdater, and the ConstantUpdater doesn't require global func name (because it's the default fallback option).

[the remark this answers is not in the document]
Comment (zhiics) [reply src/relay/backend/graph_runtime_codegen.cc:379]: I am saying that we probably don't want to duplicate the code in both compiler.cc and build_module.cc. Instead, we can move it into ConstantUpdater so that difference places can share the same code. I am not saying these two lines are not needed.

[the remark this answers is not in the document]
Comment (mbaret) [reply src/relay/backend/graph_runtime_codegen.cc:379]: Agreed, but there's about 20 shared lines in this change between the two, not just those two. That's why I suggested creating a shared utility function rather than moving it all into ConstantUpdater.

[the remark this answers is not in the document]
Comment (zhiics) [reply src/relay/backend/graph_runtime_codegen.cc:379]: I am fine with that as well

**Comment by manupa-arm** [comment]
LGTM

**Review by manupa-arm — APPROVED** [review]

**Review by leandron — APPROVED** [review]
It looks like there some CI issue within the validation of this PR, maybe just need to re trigger it.
FWIW, I successfully tested this locally on my machine to generate Ethos-N compiled modules.

**Review by zhiics — APPROVED** [review]
